docs: add concurent compaction docs (#15218)
apache/druid: docs/data-management/automatic-compaction.md

}
```

## Concurrent append and replace

:::info
Concurrent append and replace is an [experimental feature](../development/experimental.md) and is not currently available for SQL-based ingestion.
:::

This feature allows you to safely replace the existing data in an interval of a datasource while new data is being appended to that interval. One of the most common applications of this is appending new data (using say streaming ingestion) to an interval while compaction of that interval is already in progress.

To set up concurrent append and replace, you need to ensure that your ingestion jobs have the appropriate lock types:

You can enable concurrent append and replace by ensuring the following:
- The append task (with `appendToExisting` set to `true`) has `taskLockType` set to `APPEND` in the task context.
- The replace task (with `appendToExisting` set to `false`) has `taskLockType` set to `REPLACE` in the task context.
- The segment granularity of the append task is equal to or finer than the segment granularity of the replace task.

> :::info

Author: renders as 

![image](https://github.com/apache/druid/assets/53799971/789f2856-7001-4378-9791-fcbb91a6b90f)

When using concurrent append and replace, keep the following in mind:

> - When the job with an `APPEND` task has a coarser granularity than the job with a `REPLACE` task, you shouldn't use concurrent append and replace. For example, if the `APPEND` job has a yearly granularity and the `REPLACE` job has a monthly granularity. The job that finishes second will fail. 

Reviewer: ```suggestion
- Concurrent append and replace fails if the task with `APPEND` lock uses a coarser segment granularity than the task with the `REPLACE` lock. For example, if the `APPEND` task uses a segment granularity of YEAR and the `REPLACE` task uses a segment granularity of MONTH, you should not use concurrent append and replace.
```

}
```

## Concurrent compaction

:::info

> Concurrent compaction is an [experimental feature](../development/experimental.md) and is not currently available for SQL-based ingestion.

Author: @AmatyaAvadhanula will concurrent compaction w/ MSQ be in for 28 or no?

Author: Answered elsewhere. It won't be.

- If the job with an `APPEND` task and the job with a `REPLACE` task have the same segment granularity. For example, when a datasource and its streaming ingestion job have the same granularity.
- If the job with an `APPEND` task  has a finer segment granularity than the replacing job.

We do not recommend using concurrent compaction when the job with an `APPEND` task has a coarser granularity than the job with a `REPLACE` task. For example, if the `APPEND` job has a yearly granularity and the `REPLACE` job has a monthly granularity. The job that finishes second will fail.
 
### Configure concurrent compaction

##### Update the compaction settings with the API
 
 First, prepare your datasource for concurrent compaction by setting its task lock type to `REPLACE`.
Add the `taskContext` like you would any other auto-compaction setting through the API:

```shell
curl --location --request POST 'http://localhost:8081/druid/coordinator/v1/config/compaction' \
--header 'Content-Type: application/json' \
--data-raw '{
    "dataSource": "YOUR_DATASOURCE",
    "taskContext": {
        "taskLockType": "REPLACE"
    }
}'
```

##### Update the compaction settings with the UI

> In the **Compaction config** for a datasource, set  **Allow concurrent compaction append tasks** to **True**.

Reviewer 2: The UI for compaction says "Allow concurrent compactions (experimental)"
<img width="563" alt="image" src="https://github.com/apache/druid/assets/95239065/7e162ef7-35cc-4704-bbe8-9eb1a2a8e22f">

> In the **Compaction config** for a datasource, set  **Allow concurrent compaction append tasks** to **True**.

Author: ```suggestion
In the **Compaction config** for a datasource, set  **Allow concurrent compactions (experimental)** to **True**.
```

##### Add the task lock type through the API

Add the following JSON snippet to your supervisor or ingestion spec if you're using the API:

   ```json
   "context": {
      "taskLockType": LOCK_TYPE
   }   
   ```
 
The `LOCK_TYPE` depends on what you're trying to accomplish.

Set  `taskLockType` to `REPLACE` if you're replacing data. For example, if you use any of the following partitioning types, use `REPLACE`:

- hash partitioning 
- range partitioning
- dynamic partitioning with append to existing set to `false`

Set `taskLockType` to  `APPEND` if dynamic partitioning with append to existing is set to `true`. 

If you have multiple append jobs all targeting the same datasource and want them to run simultaneously, you need to also include the following context parameter:

```json
"useSharedLock": "true"
```

Keep in mind that `taskLockType` takes precedence over `useSharedLock`. Do not use it with `REPLACE` task locks.

##### Add a task lock using the Druid console

> As part of the batch or streaming ingestion **Load data** wizards, you can, you can choose **Allow concurrent append tasks** to use concurrent compaction. 

Author: ```suggestion
As part of the batch or streaming ingestion **Load data** wizards, you can choose **Allow concurrent append tasks** to use concurrent compaction. 
```

}
```

## Concurrent append and replace

:::info
Concurrent append and replace is an [experimental feature](../development/experimental.md) and is not currently available for SQL-based ingestion.
:::

> If you enable automatic compaction, you can use concurrent append and replace to concurrently compact data as you ingest it for streaming and legacy JSON-based batch ingestion. 

Reviewer: ```suggestion
This feature allows you to safely replace the existing data in an interval of a datasource while new data is being appended to that interval. One of the most common applications of this is appending new data (using say streaming ingestion) to an interval while compaction of that interval is already in progress.
```

> Setting up concurrent append and replace is a two-step process. The first is to update your datasource and the second is to update your ingestion job.
> 
> Using concurrent append and replace in the following scenarios can be beneficial:
> 
> - If the job with an `APPEND` task and the job with a `REPLACE` task have the same segment granularity. For example, when a datasource and its streaming ingestion job have the same granularity.
> - If the job with an `APPEND` task  has a finer segment granularity than the replacing job.

Reviewer: ```suggestion
You can enable concurrent append and replace by ensuring the following:
- The append task (with `appendToExisting` set to `true`) has `taskLockType` set to `APPEND` in the task context.
- The replace task (with `appendToExisting` set to `false`) has `taskLockType` set to `REPLACE` in the task context.
- The segment granularity of the append task is equal to or finer than the segment granularity of the replace task.
```

> We do not recommend using concurrent append and replace when the job with an `APPEND` task has a coarser granularity than the job with a `REPLACE` task. For example, if the `APPEND` job has a yearly granularity and the `REPLACE` job has a monthly granularity. The job that finishes second will fail.

Reviewer: This point should be in a note or warning block.

Two more points to call out are that:

```
At any point in time
- There can only be a single task that holds a `REPLACE` lock on a given interval of a datasource.
- There may be multiple tasks that hold `APPEND` locks on a given interval of a datasource and append data to that interval simultaneously.
```

 
### Configure concurrent append and replace

##### Update the compaction settings with the API
 
 First, prepare your datasource for concurrent append and replace by setting its task lock type to `REPLACE`.
Add the `taskContext` like you would any other auto-compaction setting through the API:

```shell
curl --location --request POST 'http://localhost:8081/druid/coordinator/v1/config/compaction' \
--header 'Content-Type: application/json' \
--data-raw '{
    "dataSource": "YOUR_DATASOURCE",
    "taskContext": {
        "taskLockType": "REPLACE"
    }
}'
```

##### Update the compaction settings with the UI

In the **Compaction config** for a datasource, set  **Allow concurrent compactions (experimental)** to **True**.

#### Add a task lock type to your ingestion job

Next, you need to configure the task lock type for your ingestion job. For streaming jobs, the context parameter goes in your supervisor spec. For legacy JSON-based batch ingestion, the context parameter goes in your ingestion spec. You can provide the context parameter through the API like any other parameter for a streaming ingestion or JSON-based batch ingestion or UI.

> ##### Add the task lock type through the API

Reviewer 2: Could we please explicitly add that a supervisor spec must always have an APPEND lock when using concurrent append and replace?

Author: Added below in the append section that talks about lock type
